pandas/core/internals/blocks.py

```diff
@@ -1000,8 +999,8 @@ def putmask(self, mask, new, align=True, inplace=False, axis=0,
                 if not (mask.shape[-1] == len(new) or
                         mask[mask].shape[-1] == len(new) or
                         len(new) == 1):
-                    raise ValueError("cannot assign mismatch "
```

Comment:
does this need to change? (or is this what is actually causing the issue); this is not exposed to the user, correct?

Reply:
Yes this ValueError caused Block.replace and ObjectBlock.replace to form infinite loop and ultimately raise RecursionError. It's because Block.replace catches exception include ValueError and delegate replace to ObjectBlock if fails. But ObjectBlock.replace also uses super(ObjectBlock, self), hence the loop. The exception handling includes TypeError and ValueError and they are both dependent on (as appear to me) by other casting cases.

Open to suggestion, not sure a better way to handle this but at the moment this change is needed to address the issue in question.

This can be exposed to user when:
```
import pandas as pd

s = pd.Series(range(10))
mask = s > 5

s[mask] = [5, 4, 3, 2, 1]
```

Reply:
do you know why we are catching ValueError at a higher level, maybe can get away with only catching TypeError and let ValueError buble up?

Reply:
Good point and it appears below test (only this test) fails when we only catch TypeError:

=================================== FAILURES ===================================
_________________ TestDataFrameReplace.test_replace_datetimetz _________________
[gw0] darwin -- Python 3.5.5 /Users/mingli/.virtualenvs/pd-dev/bin/python3.5

self = <pandas.tests.frame.test_replace.TestDataFrameReplace object at 0x1112dc400>

    def test_replace_datetimetz(self):
    
        # GH 11326
        # behaving poorly when presented with a datetime64[ns, tz]
        df = DataFrame({'A': date_range('20130101', periods=3,
                                        tz='US/Eastern'),
                        'B': [0, np.nan, 2]})
        result = df.replace(np.nan, 1)
        expected = DataFrame({'A': date_range('20130101', periods=3,
                                              tz='US/Eastern'),
                              'B': Series([0, 1, 2], dtype='float64')})
        assert_frame_equal(result, expected)
    
        result = df.fillna(1)
        assert_frame_equal(result, expected)
    
        result = df.replace(0, np.nan)
        expected = DataFrame({'A': date_range('20130101', periods=3,
                                              tz='US/Eastern'),
                              'B': [np.nan, np.nan, 2]})
        assert_frame_equal(result, expected)
    
        result = df.replace(Timestamp('20130102', tz='US/Eastern'),
                            Timestamp('20130104', tz='US/Eastern'))
        expected = DataFrame({'A': [Timestamp('20130101', tz='US/Eastern'),
                                    Timestamp('20130104', tz='US/Eastern'),
                                    Timestamp('20130103', tz='US/Eastern')],
                              'B': [0, np.nan, 2]})
        assert_frame_equal(result, expected)
    
        result = df.copy()
        result.iloc[1, 0] = np.nan
        result = result.replace(
            {'A': pd.NaT}, Timestamp('20130104', tz='US/Eastern'))
        assert_frame_equal(result, expected)
    
        # coerce to object
        result = df.copy()
        result.iloc[1, 0] = np.nan
        result = result.replace(
>           {'A': pd.NaT}, Timestamp('20130104', tz='US/Pacific'))

pandas/tests/frame/test_replace.py:1060: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/core/frame.py:3895: in replace
    method=method)
pandas/core/generic.py:6016: in replace
    convert=convert)
pandas/core/internals/managers.py:560: in replace
    return self.apply('replace', **kwargs)
pandas/core/internals/managers.py:421: in apply
    applied = getattr(b, f)(**kwargs)
pandas/core/internals/blocks.py:798: in replace
    blocks = self.putmask(mask, value, inplace=inplace)
pandas/core/internals/blocks.py:1800: in putmask
    new_values, _, new, _ = self._try_coerce_args(new_values, new)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = DatetimeTZBlock: slice(0, 1, 1), 1 x 3, dtype: datetime64[ns, US/Eastern]
values = array([[ 1357016400000000000, -9223372036854775808,  1357189200000000000]])
other = Timestamp('2013-01-04 00:00:00-0800', tz='US/Pacific')

    def _try_coerce_args(self, values, other):
        """
            localize and return i8 for the values
    
            Parameters
            ----------
            values : ndarray-like
            other : ndarray-like or scalar
    
            Returns
            -------
            base-type values, values mask, base-type other, other mask
            """
        values_mask = _block_shape(isna(values), ndim=self.ndim)
        # asi8 is a view, needs copy
        values = _block_shape(values.asi8, ndim=self.ndim)
        other_mask = False
    
        if isinstance(other, ABCSeries):
            other = self._holder(other)
            other_mask = isna(other)
    
        if isinstance(other, bool):
            raise TypeError
        elif (is_null_datelike_scalar(other) or
              (lib.is_scalar(other) and isna(other))):
            other = tslibs.iNaT
            other_mask = True
        elif isinstance(other, self._holder):
            if other.tz != self.values.tz:
                raise ValueError("incompatible or non tz-aware value")
            other_mask = _block_shape(isna(other), ndim=self.ndim)
            other = _block_shape(other.asi8, ndim=self.ndim)
        elif isinstance(other, (np.datetime64, datetime, date)):
            other = tslibs.Timestamp(other)
            tz = getattr(other, 'tz', None)
    
            # test we can have an equal time zone
            if tz is None or str(tz) != str(self.values.tz):
>               raise ValueError("incompatible or non tz-aware value")
E               ValueError: incompatible or non tz-aware value

pandas/core/internals/blocks.py:2898: ValueError

Reply:
@jreback made some changes now we don't touch above error but instead make sure we don't enter infinite loop.